### PR TITLE
[Merged by Bors] - update archetypes if needed before running system in SingleThreadedExecutor

### LINKED ISSUE #1585: Shared query iteration progress between stages?
## Bevy version

f61e44db2824c0cc4b4815748c4d3f051002ada7 (latest main right now)

## Operating system & version

Arch Linux, Kernel 5.11.2-zen1-1-zen

## What you did

```rust
use bevy::prelude::*;

struct Component;

fn main() {
    App::build()
        .add_plugins(DefaultPlugins)
        .add_startup_system(setup.system())
        .add_system(insert_component_system.system().label("send"))
        .add_stage_after(CoreStage::Update, "CustomStage", SystemStage::single(count_entities.system()))
        // this works fine, though: .add_system(count_entities.system().label("receive").after("send"))
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn(()).spawn(()).spawn(());
}

fn insert_component_system(
    keyboard: Res<Input<KeyCode>>,
    query: Query<Entity>,
    mut commands: Commands,
) {
    if keyboard.just_pressed(KeyCode::Return) {
        for entity in query.iter() {
            // without this line it seems to work fine
            commands.insert(entity, Component);

            break;
        }
    }
}

fn count_entities(query: Query<Entity>) {
    println!("entities: {}", query.iter().len());
}
```

## What you expected to happen

Even after pressing the return key, it should print out "entities: 3".

## What actually happened

After each press of the return key, the number decreased by one (eg. "entities: 2", ...)

## Additional information

Moving the counting system into the same stage as the modifying system seems to resolve the issue.

Since we're not iterating through all entities, could a possible cause be that the query iteration progress is shared between queries of different stages?

This is the minimal reproducing code I could construct. The original code worked fine on 0.4, but while trying to upgrade to `main` this came up.

An alternative repro is the following, without the `break`:
```rust
use bevy::prelude::*;

struct ComponentA;
struct ComponentB;

fn main() {
    App::build()
        .add_plugins(DefaultPlugins)
        .add_startup_system(setup.system())
        .add_system(insert_component_system.system().label("send"))
        .add_stage_after(CoreStage::Update, "CustomStage", SystemStage::single(count_entities.system()))
        // this works fine, though: .add_system(count_entities.system().label("receive").after("send"))
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn(()).with(ComponentA).spawn(()).spawn(());
}

fn insert_component_system(
    keyboard: Res<Input<KeyCode>>,
    query: Query<Entity, With<ComponentA>>,
    mut commands: Commands,
) {
    if keyboard.just_pressed(KeyCode::Return) {
        for entity in query.iter() {
            commands.insert(entity, ComponentB);
        }
    }
}

fn count_entities(query: Query<Entity>) {
    println!("entities: {}", query.iter().len());
}
```

### PR DESCRIPTION
**Pull request by mockersf**
fixes #1585 

I copied most of the logic from the `ParallelSystemExecutor` impl, simplifying it a little as systems can't run in parallel

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/executor.rs [hunk 1]
```diff
@@ -10,17 +10,57 @@ pub trait ParallelSystemExecutor: Downcast + Send + Sync {
 
 impl_downcast!(ParallelSystemExecutor);
 
-#[derive(Default)]
-pub struct SingleThreadedExecutor;
+// #[derive(Default)]
```
Comment (Ratysz) [inline crates/bevy_ecs/src/schedule/executor.rs:13]:
```suggestion
```

**Review by Ratysz — COMMENTED** [review]
Ideally, this should come with a new test in stage.rs. They all cover both single-threaded and parallel executors, but there isn't one that verifies if archetype changes are properly handled.

**Comment by mockersf** [comment]
added tests, it fails without this fix

**Review by Ratysz — APPROVED** [review]
I would've made those a single test, for code reuse and consistency with the others. Other than that, looks good to me.

**Comment by cart** [comment]
Looks good to me. Thanks for the quick fix! I saw the discord discussion and was worried we had something more nasty on our hands :smile:

Im cool with leaving the tests separate here as the boilerplate isn't huge and it ensures they are completely isolated from each other.

**Pull request closed without merging**